scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -272,6 +274,35 @@
          (catch Exception e
            (log/error e "Mesos scheduler status update error")))))
 
+(defn handle-framework-message
+  "Processes a framework message from Mesos."
+  [conn framework-message]
+  (timers/time!
+    handle-framework-message-duration
+    (try
+      (let [db (db conn)
+            {:strs [exit-code progress-message progress-percent sandbox-directory task-id] :as message}
+            (-> (String. ^bytes framework-message "UTF-8") (json/read-str))
+            _ (log/debug "Received framework message:" {:task-id task-id, :message message})
+            _ (when (str/blank? task-id)
+                (throw (ex-info "task-id is empty in framework message" {:message message})))
+            instance (ffirst (q '[:find ?i
+                                  :in $ ?task-id
+                                  :where [?i :instance/task-id ?task-id]]
+                                db task-id))]
+        (if (nil? instance)
+          (throw (ex-info "No instance found!" {:task-id task-id}))
+          (let [txns (cond-> []
+                              exit-code (conj [:db/add instance :instance/exit-code (int exit-code)])
+                              progress-message (conj [:db/add instance :instance/progress-message (str progress-message)])
+                              progress-percent (conj [:db/add instance :instance/progress (int progress-percent)])
+                              sandbox-directory (conj [:db/add instance :instance/sandbox-directory (str sandbox-directory)]))]
+            (when (seq txns)
+              (log/info "Updating instance" instance "to" txns)
+              (transact-with-retries conn txns)))))
+      (catch Exception e
+        (log/error e "Mesos scheduler framework message error")))))
```

Comment:
The error should be a rare scenario, e.g. if the transaction failed. We do not expect to have issues with parsing and validating the json data.